pandas/core/dtypes/cast.py

```diff
@@ -159,29 +168,19 @@ def maybe_downcast_to_dtype(result, dtype):
         return converted
 
     # a datetimelike
-    # GH12821, iNaT is casted to float
-    if dtype.kind in ["M", "m"] and result.dtype.kind in ["i", "f"]:
-        if hasattr(dtype, "tz"):
+    # GH12821, iNaT is cast to float
+    if is_datetime_or_timedelta_any_dtype(dtype) and result.dtype.kind in ["i", "f"]:
```

Comment:
the dtype.kind checks are much more performant

Reply:
i agree, just leave this as is i think for L163 (and can remove the comment which isold)

Reply:
Ok! Done both